instant-distance-py/src/lib.rs

```diff
@@ -414,73 +413,42 @@ impl Neighbor {
     }
 }
 
-#[repr(align(32))]
 #[derive(Clone, Deserialize, Serialize)]
-struct FloatArray(#[serde(with = "BigArray")] [f32; DIMENSIONS]);
+struct FloatArray<M> {
+    array: Vec<f32>,
+    phantom: PhantomData<M>,
+}
 
-impl FloatArray {
+impl<M: Metric> FloatArray<M> {
     fn try_from_pylist(list: &PyList) -> Result<Vec<Self>, PyErr> {
         list.into_iter().map(FloatArray::try_from).collect()
     }
 }
 
-impl TryFrom<&PyAny> for FloatArray {
+impl<M: Metric> From<Vec<f32>> for FloatArray<M> {
+    fn from(array: Vec<f32>) -> Self {
+        Self {
+            array,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<M: Metric> TryFrom<&PyAny> for FloatArray<M> {
     type Error = PyErr;
 
     fn try_from(value: &PyAny) -> Result<Self, Self::Error> {
-        let mut new = FloatArray([0.0; DIMENSIONS]);
-        for (i, val) in value.iter()?.enumerate() {
-            match i >= DIMENSIONS {
-                true => return Err(PyTypeError::new_err("point array too long")),
-                false => new.0[i] = val?.extract::<f32>()?,
-            }
-        }
-        Ok(new)
+        let array: Vec<f32> = value
+            .iter()?
+            .map(|val| val.and_then(|v| v.extract::<f32>()))
+            .collect::<Result<_, _>>()?;
+        Ok(Self::from(array))
```

Comment:
If you ditch the `array` binding in favor of yielding `Self::from(...collect())` directly, you don't need the type annotation.

Reply:
There might be too many steps for rustc to figure this out. I still need type annotation on `collect`:
```
    Ok(Self::from(
            value
                .iter()?
                .map(|val| val.and_then(|v| v.extract::<f32>()))
                .collect::<Result<Vec<f32>, _>>()?,
        ))
```
If I remove even `Vec<f32>` part then it fails to compile:
```
error[E0277]: the trait bound `FloatVector<M>: From<()>` is not satisfied
   --> instant-distance-py/src/lib.rs:431:13
    |
430 |           Ok(Self::from(
    |              ---------- required by a bound introduced by this call
431 | /             value
432 | |                 .iter()?
433 | |                 .map(|val| val.and_then(|v| v.extract::<f32>()))
434 | |                 .collect::<Result<_, _>>()?,
    | |___________________________________________^ the trait `From<()>` is not implemented for `FloatVector<M>`
    |
    = help: the trait `From<Vec<f32>>` is implemented for `FloatVector<M>`

error[E0277]: a value of type `()` cannot be built from an iterator over elements of type `f32`
    --> instant-distance-py/src/lib.rs:431:13
     |
431  | /             value
432  | |                 .iter()?
433  | |                 .map(|val| val.and_then(|v| v.extract::<f32>()))
     | |________________________________________________________________^ value of type `()` cannot be built from `std::iter::Iterator<Item=f32>`
434  |                   .collect::<Result<_, _>>()?,
     |                    ------- required by a bound introduced by this call
     |
     = help: the trait `FromIterator<f32>` is not implemented for `()`
     = help: the trait `FromIterator<()>` is implemented for `()`
     = note: required for `Result<(), PyErr>` to implement `FromIterator<Result<f32, PyErr>>`
note: required by a bound in `collect`
    --> /home/kuba/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1832:19
     |
1832 |     fn collect<B: FromIterator<Self::Item>>(self) -> B
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `collect`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `instant-distance-py` due to 2 previous errors
```

I get the same error if I remove the whole type annotation.